urllib3/util/rfc6555.py

```diff
@@ -0,0 +1,223 @@
+_ASYNC_ERROR_NUMBERS = set([errno.EINPROGRESS,
+                            errno.EAGAIN,
+                            errno.EWOULDBLOCK])
+if hasattr(errno, "WSAEWOULDBLOCK"):  # Platform-specific: Windows
+    _ASYNC_ERROR_NUMBERS.add(errno.WSAEWOULDBLOCK)
+
+_IP_FAMILIES = set([socket.AF_INET])
+if hasattr(socket, "AF_INET6"):
+    _IP_FAMILIES.add(socket.AF_INET6)
+
+
+class _HappyEyeballs(object):
+    def __init__(self, address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
+                 source_address=None, socket_options=None):
+        self._address = address
+        self._error = None
+        self._timeout = timeout
+        self._source_address = source_address
+        self._socket_options = socket_options
+        self._sockets = []
+        self._selector = DefaultSelector()
+        self._start_time = None
+
+    def connect(self):
+        """ Given the setup of the algorithm, attempt to
+        use a cached value and if it doesn't work then try
+        to use a """
+        self._start_time = current_time()
+        try:
+            sock = None
+            if self._address in _HAPPY_EYEBALLS_CACHE:
+                family, expires = _HAPPY_EYEBALLS_CACHE[self._address]
+
+                # Invalidate the cache if it's expired.
+                if expires > current_time():
+                    del _HAPPY_EYEBALLS_CACHE[self._address]
+                else:
+                    try:
+                        sock = self._algorithm(family)
+                    # Invalidate the cache if connecting doesn't work.
+                    except (OSError, socket.error):
+                        del _HAPPY_EYEBALLS_CACHE[self._address]
+
+            if sock is None:
+                sock = self._algorithm()
+
+                # Add a new entry to the cache if successful.
+                cache_entry = (sock.family,
+                               current_time() + _HAPPY_EYEBALLS_CACHE_TIME)
+                _HAPPY_EYEBALLS_CACHE[self._address] = cache_entry
+
+            return sock
+        finally:
+            self._cleanup()
+
+    def _algorithm(self, family=None):
+        """ Executes the algorithm for a given family
+        or no family if IPv4 or IPv6 are allowed. """
+        if family is None:
+            family = socket.AF_UNSPEC
+        host, port = self._address
+        dns_results = socket.getaddrinfo(host, port, family, socket.SOCK_STREAM)
+        dns_results = list(filter(lambda *args: args[0] not in _IP_FAMILIES,
+                                  dns_results))
+
+        result = None
+        for family, socktype, proto, _, sa in dns_results:
+            self._create_socket(family, socktype, proto, sa)
+            result = self._wait_for_connection(False)
+            if result:
+                break
+            self._timeout_check()
+
+        if result is None:
+            if self._error:
+                raise self._error
+            else:
+                raise socket.error(errno.ECONNREFUSED)
```

Comment:
So this code isn't actually unreachable, there's at least one scenario I can think of. The call to `getaddrinfo` returns only non-IPv4 and non-IPv6 names there will be no `gaierror` and no sockets for the algorithm to try so no errors and then it'll raise this error because there's also no result.
That error should probably happen earlier, like right after the dns_results are filtered.